dbt/clients/jinja.py

```diff
@@ -194,17 +194,37 @@ class ParserMacroCapture(jinja2.Undefined):
         def __init__(self, hint=None, obj=None, name=None,
                      exc=None):
             super(jinja2.Undefined, self).__init__()
-
             self.node = node
             self.name = name
             self.package_name = node.get('package_name')
-
-        def __getattr__(self, name):
-
             # jinja uses these for safety, so we have to override them.
             # see https://github.com/pallets/jinja/blob/master/jinja2/sandbox.py#L332-L339 # noqa
-            if name in ['unsafe_callable', 'alters_data']:
-                return False
+            self.unsafe_callable = False
+            self.alters_data = False
+
+        def __deepcopy__(self, memo):
+            path = os.path.join(self.node.get('root_path'),
+                                self.node.get('original_file_path'))
+
+            logger.debug(
+                'A ParserMacroCapture has been deecopy()d, invalid reference '
+                'to "{}" in node {}.{} (source path: {})'
+                .format(self.name, self.node.get('package_name'),
+                        self.node.get('name'),
+                        path))
+
+            dbt.exceptions.raise_compiler_error(
+                'dbt has detected at least one invalid reference in {}.{}. '
+                'Check logs for more information'
+                .format(self.node.get('package_name'), self.node.get('name'))
+            )
+
+        def __getattr__(self, name):
+            if name == 'name' or name.startswith('__') and name.endswith('__'):
+                raise AttributeError(
```

Comment:
why would we raise this error for `__*__`?

Reply:
because `copy.deepcopy` does a series of successive `hasattr(x, '__*__')` checks to decide how to deepcopy things, and `hasattr(obj, name)` is basically implemented as:

```
try:
  getattr(obj, name)
except AttributeError:
  return False
else:
  return True
```

Since `__getattr__` is only accessed when an attribute is not already defined, we want to signal that we don't implement those methods.

Reply:
roger!